ext/ffi/fast_call.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+pub(crate) fn compile_trampoline(sym: &Symbol) -> Trampoline {
+  Trampoline::compile(sym)
+}
+
+pub(crate) fn make_template(sym: &Symbol, trampoline: &Trampoline) -> Template {
+  let args = once(fast_api::Type::V8Value)
+    .chain(sym.parameter_types.iter().map(|t| t.into()))
+    .collect::<Vec<_>>();
+
+  Template {
+    args: args.into_boxed_slice(),
+    ret: (&fast_api::Type::from(&sym.result_type)).into(),
+    symbol_ptr: trampoline.ptr(),
+  }
+}
+
+/// Trampoline for fast-call FFI functions
+///
+/// Removes first argument (Javascript object receiver) and shifts the rest of arguments to the left
+pub(crate) struct Trampoline(ExecutableBuffer);
+
+impl Trampoline {
+  fn ptr(&self) -> *const c_void {
+    &self.0[0] as *const u8 as *const c_void
+  }
+
+  fn compile(sym: &Symbol) -> Self {
+    #[derive(Clone, Copy)]
+    struct ArgCounter {
+      integer: i32,
+      sse: i32,
+    }
+    let mut counter = ArgCounter {
+      integer: -6,
+      sse: -8,
+    };
+    let args_to_shift: Vec<_> = sym
+      .parameter_types
+      .iter()
+      .filter_map(|p| match p {
+        NativeType::F32 | NativeType::F64 => {
+          counter.sse += 1;
+          if counter.sse > 0 && counter.integer >= 0 {
+            // floats are only shifted to accomodate integer shift in the stack
+            Some(counter)
+          } else {
+            None
+          }
+        }
+        _ => {
+          counter.integer += 1;
+          Some(counter)
+        }
+      })
+      .collect();
+
+    let mut ops = dynasmrt::x64::Assembler::new().unwrap();
+    if args_to_shift.is_empty() {
+      dynasm!(ops
+        ; .arch x64
+        ; xor rdi, rdi
+      );
+    } else {
+      for arg in args_to_shift {
+        match arg {
+          ArgCounter { integer: -5, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdi, rsi
+          ),
+          ArgCounter { integer: -4, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rsi, rdx
+          ),
+          ArgCounter { integer: -3, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rdx, rcx
+          ),
+          ArgCounter { integer: -2, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov rcx, r8
+          ),
+          ArgCounter { integer: -1, .. } => dynasm!(ops
+            ; .arch x64
+            ; mov r8, r9
+          ),
+          ArgCounter { integer: 0, sse } => dynasm!(ops
+            ; .arch x64
+            ; mov r9, [rsp + (1 + sse.max(0)) * 8]
+          ),
+          ArgCounter { integer, sse } => dynasm!(ops
+            ; .arch x64
+            ; mov rax, [rsp + (1 + integer + sse.max(0)) * 8]
+            ; mov [rsp + (integer + sse.max(0)) * 8], rax
+          ),
+        }
+      }
+    }
+
+    // tail-call
+    // SAFETY: stack pointer is not modified, therefore stack remains (un)aligned.
+    // Stack contains one extra parameter (the last one) when using >6 parameters, which the function will ignore
```

Comment:
"when using >6 **integer** parameters"

Reply:
Why not write it in?
```suggestion
    // Stack contains one extra parameter (the last one) when using >6 integer parameters, which the function will ignore
```